### PR TITLE
New pseudo-keys:  Part I

### PR DESCRIPTION
**Pull request by bpoldrack**
Sits on top of PR #707 and resolves its issue WRT `is_asset_directory` by removing that key altogether. This is now split into the new `onyo.is.asset` and `onyo.is.directory` pseudo-keys. Although for the most part `onyo.is.directory` is what is used instead.

This is the first PR of a series, aiming to implement #662, #688, #689.
It's introducing an `Item` class and a new `PSEUDO_KEYS` mapping, that add pseudo-keys to dict-likes.
The aim for this PR is:
- Make `path` and `directory` keys aliases and replace their internal usage (except for test code, where they are still used at times in their role as default aliases)
- Remove `is_asset_directory` key in favor of `onyo.is.directory` (and `onyo.is.asset`)
- Replace `DotNotationWrapper` by `Item` in most places. `DotNotationWrapper` remains valid in some places where we don't want any pseudo-keys yet (For example: Reading yaml from an edited temporary file - only becomes an `Item` when we tie it to the actual repo. We don't want pseudo-keys referring to the temp file)
- Don't break anything other than `is_asset_directory` while throwing this class in.
- It does not yet utilize `Item`'s ability to represent directories, templates, etc. - just assets in the first round.
- It does not yet implement git metadata keys, but builds on a mechanic that is meant to account for the need to lazy load pseudo-keys.
- It does not yet introduce a proper mechanic for namespaces nor streamline the usage of this class yet. I want to see what this really needs first.

The scope is limited, in order to address that topic in digestable chunks. Next up should be:
- git metadata
- actually use `Item` consistently for anything at high-level code, so this becomes the datastructure, that `Inventory` and command implementations deal with, instead of arbitrary dicts and paths.
- Build different validators (at least in terms of design). Anything is an Item, but what it's used for is what makes the difference. If I want to use an asset as a template, I simply treat it that way (and put it into a TemplateValidator). As opposed to the initial idea of the object itself determining what it is. There may be more to do in a similar fashion along this line of thought.
- consequently properly implement namespaces and allow `get` to query anything (dirs, templates)

During these further steps, we/I can figure out additional functions to ease various ways of dealing with these objects (we need to repeatedly strip or reset pseudo-keys and we have notions of "you-can't-set-these", leading to a bunch of code duplication - I want to iron that out as we get to see more access patterns)

### CODE REVIEW

File: onyo/lib/items.py
Comment (aqw) [inline onyo/lib/items.py:108]: Does not need to be addressed in this PR, but I think I'd prefer these to all be private methods.

They're used internally to populate the pseudokeys, but everything else uses the pseudokey namespace.

File: onyo/lib/items.py
Comment (aqw) [inline onyo/lib/items.py:163]: You read my mind. This was also something I was hoping the new object would allow us to address.

File: onyo/lib/pseudokeys.py
Comment (aqw) [inline onyo/lib/pseudokeys.py:47]: Nit:
> Whether the object is an asset.
to
> Is the item an asset.

The purpose is to consolidate vocabulary. "Whether" is just a semi-synonym for "is", and "object" for "item".

The pseudokey vocabulary was well chosen. We can/should stick with it.

File: onyo/lib/pseudokeys.py
Comment (aqw) [inline onyo/lib/pseudokeys.py:62]: This, in the near future will change. IMO, the entire `onyo.` namespace should be reserved (and nothing else).

File: onyo/lib/tests/test_commands_get.py
Comment (aqw) [inline onyo/lib/tests/test_commands_get.py:106]: Oh ugh. I remember this.

Something we will be sure to unroll by the end of all of this pseudo-key work.

File: onyo/tests/demo/reference_git_log.txt
Comment (aqw) [inline onyo/tests/demo/reference_git_log.txt:138]: Looks like a bug snuck in.

Is this known and will be fixed in the upcoming PRs?

**Review by aqw — APPROVED** [review]
Overall LGTM!

It's a huge PR, so it's difficult to review in depth.

In general, I chose to not be nitpicky, as this is all clearly heading in a positive direction, and it cannot fix everything on its way there. There will be some mess.

With that in mind, I saw no showstoppers.

Please review my comments. If you want to fix any of them in this PR, you are welcome to, but it's not necessary. They can be addressed in later PRs as more pieces land.

File: onyo/cli/get.py
Comment (bpoldrack) [reply onyo/cli/get.py:164]: Yeah, I'm fine with that. Config cache + "real" aliases first, though.

File: onyo/lib/commands.py
Comment (bpoldrack) [reply onyo/lib/commands.py:299]: True. Missed that spot. Will fix in this PR.

File: onyo/lib/commands.py
Comment (bpoldrack) [reply onyo/lib/commands.py:253]: Agree.

Comment (bpoldrack) [reply onyo/lib/commands.py:253]: Agree.

File: onyo/lib/items.py
Comment (bpoldrack) [reply onyo/lib/items.py:163]: Yeah, just wanted to see to what extent DotNotationWrapper or even regular dicts keep a role first (where something like that helper may be needed). In any case: At least part if not all of that helper will be simplified and go into `DotNotationWrapper/Item.__eq__`.

File: onyo/lib/pseudokeys.py
Comment (bpoldrack) [reply onyo/lib/pseudokeys.py:47]: Agree. Will adjust.

File: onyo/lib/pseudokeys.py
Comment (bpoldrack) [reply onyo/lib/pseudokeys.py:62]: I guess, aliases are going to be a dynamic part of what is reserved, though. However, that may be a thing for fsck only.

File: onyo/tests/demo/reference_git_log.txt
Comment (bpoldrack) [reply onyo/tests/demo/reference_git_log.txt:138]: Good catch, thx! Missed that one. Need to fix.

File: onyo/lib/commands.py
Comment (bpoldrack) [reply onyo/lib/commands.py:299]: Actually, no. Not needed, because the incoming dict is supposed to be an `Item`. Hence, aliases used to create it in the first place were already resolved.
The "supposed to be an Item" is key, though. Happens to be the bug you saw in the demo-repo log ;-)

File: onyo/lib/differs.py
Comment (bpoldrack) [reply onyo/lib/differs.py:40]: Agree, but I'm not changing that yet, b/c that'll be inconsistent. We currently use absolute paths and they are only sometimes coming from an asset dict. For other operations the paths are the only thing passed around.
So, this can be consistently changed in one go, when we stop passing plain paths, but `Item`s instead.

File: onyo/lib/commands.py
Comment (aqw) [inline onyo/lib/commands.py:1331]: I think this might be too strict.

I agree that our current pseudo-keys under `onyo.` cannot be unset. Especially since we decided that we won't allow creation/deletion through set/unset. Just modifying within the guardrails.

But the reserved keys:
```
RESERVED_KEYS = ['template', 'onyo'] + list(PSEUDOKEY_ALIASES.keys())
...
PSEUDOKEY_ALIASES: Dict[str, str] = {
    'path': 'onyo.path.relative',
    'directory': 'onyo.path.parent',
}
```

Any aliases should be translated before they hit this far into `onyo_unset()`. And if they somehow do get this far, I don't see the problem of unsetting keys that match.

And a similar argument for `template`. It should never exist in the asset content. But if it does somehow get into assets, I don't see why we should disallow unsetting it.

File: onyo/lib/pseudokeys.py
Comment (aqw) [inline onyo/lib/pseudokeys.py:60]: I would adjust this just a little bit:

> Pseudo-Keys are key names that are addressable but not in asset content.

to 

> Pseudo-Keys are keys that are addressable but not written to disk as YAML.

Or something like that. it actually is a bit difficult to phrase right, since `onyo.is.directory` can be written to disk (just not as YAML). And using the word "file" is correct, but gets confusing WRT asset directories.

My goal was to make it clear that it was on-disk. The asset content in-memory clearly has these keys.

File: onyo/lib/pseudokeys.py
Comment (aqw) [inline onyo/lib/pseudokeys.py:33]: Nit. And I'm not sure if it matters, but this list is not alphabeitized.

I understand they're sorted so that similar things are near each other. But we will be adding quite a few more keys. Eventually, alphabetized is going to be the only sane way to find anything.

File: onyo/lib/consts.py
Comment (aqw) [inline onyo/lib/consts.py:10]: More "talking out loud" rather than anything else.

I'm not sure if aliases should be reserved. I just think they should be translated.

Though I see this is a nice way to make sure they don't somehow leak into the asset content un-translated.

I'm split.

Part of me wants them to just translate on the way in, and fsck does the actual checks.

File: onyo/lib/differs.py
Comment (aqw) [reply onyo/lib/differs.py:40]: Thanks for the explanation!

Can an explicit TODO be added to note this?

File: onyo/lib/inventory.py
Comment (aqw) [inline onyo/lib/inventory.py:308]: `asset['onyo.is.directory']` should always return `True` or `False`, so the `get` doesn't provide any value here, except possibly obscuring some bugs.

Something to keep in mind when the later PRs come and start using these new pseudokeys internally.

File: onyo/lib/inventory.py
Comment (aqw) [inline onyo/lib/inventory.py:310]: Nit: this comment should be in the next if statement, where it happens.

File: onyo/lib/items.py
Comment (aqw) [inline onyo/lib/items.py:32]: Interesting word "potentially".

File: onyo/lib/items.py
Comment (aqw) [inline onyo/lib/items.py:70]: This seems to do a lookup every time, with no caching.

File: onyo/lib/onyo.py
Comment (aqw) [inline onyo/lib/onyo.py:641]: This sounds like a TODO. Can it be marked as such? Otherwise we will need to read all comments to make sure they don't contain unmarked TODOs.

**Review by aqw — APPROVED** [review]
LGTM!

I took the opportunity to review the non-test code more closely, as it's my second read through.

I left comments. None are blockers. Interested in your feedback. :-)

File: onyo/cli/get.py
Comment (bpoldrack) [reply onyo/cli/get.py:165]: Yes, because this is just what it was before. The additional query abilities that this effort should ultimately lead to aren't tested yet. Also: Not sure that `file` works properly yet (probably not). 
In any case: I'd like to change/test the querying properly when the approach is applied to directories and templates that are now able to have pseudo-keys as well.

So, didn't want to put it into documentation before that's actually done and tested, even though at least `absolute` should already work.

File: onyo/lib/commands.py
Comment (bpoldrack) [reply onyo/lib/commands.py:1331]: Yes, this is just to not break stuff yet. Ultimately, the concept of reserved/disallowed needs to change (it's context-dependent) anyway. `onyo` is just a crude measure to account for the top-level key and `template` is supposed to become `onyo.cli.template` and would then not be part of the dict indeed.

So, in short: Agree, this needs to change, but only after there's a better way to specify namespaces (`onyo.` in this case) and `template` has moved to where it now belongs.

File: onyo/cli/get.py
Comment (aqw) [reply onyo/cli/get.py:165]: That makes sense. Thanks for explaining. :-)

File: onyo/lib/pseudokeys.py
Comment (bpoldrack) [reply onyo/lib/pseudokeys.py:60]: True. It's actually just taken from the previous `PSEUDO_KEYS` definition in `consts.py`.

> since onyo.is.directory can be written to disk (just not as YAML).

Yeah, and I think we said, we'd allow for several pseudo-keys to be set in templates. So, it's "on disc" and "in YAML", just not in an asset file.
So,

"Pseudo-Keys are key names that are addressable but not written to asset files."?

File: onyo/lib/inventory.py
Comment (bpoldrack) [reply onyo/lib/inventory.py:308]: Nope.  That's the ultimate values for an existing thing in the inventory, yes. But when a new asset is created as an `Item`/dict and passed to `Inventory.add_asset` this property may not be specified (at least as of now) and can't be derived from the state of things on disk, because it doesn't exist yet. It's this very function that says "well, if it wasn't specified that counts as `False`".
We may want to change that and require this to be set by `onyo_new` instead, before it's passed here. Not opposing that, it's just not (yet) what's happening.

Comment (aqw) [reply onyo/lib/inventory.py:308]: Ahh. Thanks for the explanation.

File: onyo/lib/inventory.py
Comment (bpoldrack) [reply onyo/lib/inventory.py:310]: True. Will change.

File: onyo/lib/items.py
Comment (bpoldrack) [reply onyo/lib/items.py:32]: :-D

Don't have a good phrasing yet. Point being, that this thing may be the result of `new --keys ..`, piped in YAML, merged templates etc. that are a state in memory that isn't any committed thing yet.

File: onyo/lib/items.py
Comment (bpoldrack) [reply onyo/lib/items.py:70]: Do you mean WRT aliases or the testing for whether this is still `PseudoKey`?
The former: Yes, that'll need to rely on config caching.
The latter: Would be nice to cache, but not yet sure how. That's because I'm not yet sure how to properly do the various ways of stripping/resetting/ignoring pseudo-keys in different operations/commands. These patterns will hopefully lead to proper `Item` methods, that do that. Once that's clear, we know how to cache (and how o invalidate it)

File: onyo/lib/onyo.py
Comment (bpoldrack) [reply onyo/lib/onyo.py:641]: Agree. Will do.

File: onyo/lib/pseudokeys.py
Comment (bpoldrack) [reply onyo/lib/pseudokeys.py:33]: Right. Will do. It does matter (kinda), simply because the main reason I wanted this kind of mapping in the first place, is the single entry point for humans.

File: onyo/lib/commands.py
Comment (bpoldrack) [reply onyo/lib/commands.py:254]: Thanks! I did in fact mean to only have the thing that's read from the temp file renamed. Will fix.

File: onyo/lib/items.py
Comment (aqw) [reply onyo/lib/items.py:70]: > Do you mean WRT aliases or the testing for whether this is still PseudoKey?

Neither. I mean, once we've looked up the pseudo-key value once (such as `onyo.is.directory`), we don't store it, and will execute it again if requested again.

I can understand not doing this now, and waiting until more pieces land to determine if caching such lookups is practical. But yeah, that's what I was pointing out.

File: onyo/lib/consts.py
Comment (aqw) [reply onyo/lib/consts.py:10]: > However, the way RESERVED_KEYS is used ATM, this all needs to be in there. So, this isn't meant to be the final form, but the intermediate definition that allows this PR to not break shit. We reserved these keys before this PR and after. Just not at the end of the series.

Cool. I'm fine with that then. Last time I looked seriously into this, I found just how intertwined the logic is. So +1 on doing this piece by piece.

File: onyo/lib/items.py
Comment (bpoldrack) [reply onyo/lib/items.py:70]: Have you seen line 75 below? It is stored. Not a cache exactly, but that's probably not the point.

File: onyo/lib/commands.py
Comment (bpoldrack) [reply onyo/lib/commands.py:326]: Semi-allowed at this point. I have neither fully implemented setting other pseudo-keys nor tested it. So, currently it would be settable via `--keys` but not from within the editor (like before). IOW: It's already stripped to not have it listed in the error message on setting disallowed keys, but if you set it in the edited YAML it would be ignored.

So, just a consequence of the intermediate state right now. Comment here is in-line with the code it's about. But that code will change.

File: onyo/lib/items.py
Comment (aqw) [reply onyo/lib/items.py:70]: The magic for me is in line 71. I understand it now. Thanks!

**Pull request merged**
